### PR TITLE
OneSignal Destination

### LINKED ISSUE #1452: Feature Request: OneSignal Destination
https://onesignal.com/

### PR DESCRIPTION
**Pull request by pedroslopez**
closes #1452

### CODE REVIEW

File: plugins/@grouparoo/onesignal/src/lib/export/destinationMappingOptions.ts [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+      },
+      group: {
+        singular: "OneSignal Group Tag",
+        plural: "OneSignal Group Tags",
+      },
+    },
+    properties: {
+      required: [{ key: "external_user_id", type: "string" }],
+      known: [
+        { key: "first_name", type: "string", important: true },
```
Comment (bleonard) [inline plugins/@grouparoo/onesignal/src/lib/export/destinationMappingOptions.ts:18]: Are these known tags?

File: plugins/@grouparoo/onesignal/src/lib/export/exportProfile.ts [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+
+    const value = newProfileProperties[key];
+    payload.tags[key] = formatVar(value);
+  }
+
+  // Groups are managed as tags
+  // If user is in group "High Value", the attribute "in_high_value" will be set to 1
+  for (const group of newGroups) {
+    const groupAttribute = getGroupAttribute(group);
+    payload.tags[groupAttribute] = 1;
```
Comment (bleonard) [inline plugins/@grouparoo/onesignal/src/lib/export/exportProfile.ts]: any sense if others do "true" or "false" or 1/0 or other here?

File: plugins/@grouparoo/onesignal/src/lib/export/exportProfile.ts [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+  }
+
+  for (const group of oldGroups) {
+    if (!newGroups.includes(group)) {
+      const groupAttribute = getGroupAttribute(group);
+      payload.tags[groupAttribute] = null;
+    }
+  }
+
+  try {
```
Comment (bleonard) [inline plugins/@grouparoo/onesignal/src/lib/export/exportProfile.ts:75]: For all the `newGroups` thing we should make a `Segment`?

File: plugins/@grouparoo/onesignal/src/lib/export/exportProfile.ts [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+  // Dates are stored as unix timestamps (int)
+  if (value instanceof Date) {
+    return Math.floor(value.getTime() / 1000);
+  }
+
+  return value;
+}
+
+function getGroupAttribute(groupName: string) {
+  const normalizedName = groupName.toLowerCase().replace(/ /g, "_");
```
Comment (bleonard) [inline plugins/@grouparoo/onesignal/src/lib/export/exportProfile.ts]: these names can probably have lots of other suff in them, emoticons, etc.
Should it be more like `groupName.toLowerCase().replace(/[^a-z0-9-_\/.]/g, "_");`?

File: plugins/@grouparoo/onesignal/src/lib/export/exportProfile.ts [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+  const oldKeys = Object.keys(oldProfileProperties);
+  const allKeys = new Set([...newKeys, ...oldKeys]);
+
+  if (toDelete) {
+    // When deleting, we just clear the old tags. No real deletion occurs.
+    newKeys = [];
+    newGroups = [];
+  }
+
+  for (const key of allKeys) {
```
Comment (bleonard) [inline plugins/@grouparoo/onesignal/src/lib/export/exportProfile.ts:50]: do we think keys need to be normalized too?

File: plugins/@grouparoo/onesignal/src/lib/export/destinationMappingOptions.ts (hunk 1 above)
Comment (pedroslopez) [reply plugins/@grouparoo/onesignal/src/lib/export/destinationMappingOptions.ts:18]: These tags have a special meaning in OneSignal: https://documentation.onesignal.com/docs/add-user-data-tags#user-names--preferences

image

File: plugins/@grouparoo/onesignal/src/lib/export/exportProfile.ts (hunk 2 above)
Comment (pedroslopez) [reply plugins/@grouparoo/onesignal/src/lib/export/exportProfile.ts]: Opted for 1/0 based on the OneSignal docs https://documentation.onesignal.com/docs/add-user-data-tags#recommended-tags-to-send

image

File: plugins/@grouparoo/onesignal/src/lib/export/exportProfile.ts (hunk 3 above)
Comment (pedroslopez) [reply plugins/@grouparoo/onesignal/src/lib/export/exportProfile.ts:75]: We could create Segments, but I don't see an obvious way to get an existing segment based on its filters, which is a problem when avoiding duplicates. It seems like they would have to create the segments themselves based on these tags we've added.

File: plugins/@grouparoo/onesignal/src/lib/export/exportProfile.ts (hunk 5 above)
Comment (pedroslopez) [reply plugins/@grouparoo/onesignal/src/lib/export/exportProfile.ts:50]: Yeah it seems like you can set any string as the tag key, but their docs recommend to... 

image
(https://documentation.onesignal.com/docs/add-user-data-tags#recommended-tags-to-send)

While it does work, I think normalizing to be better safe than sorry could be a good option

Comment (bleonard) [reply plugins/@grouparoo/onesignal/src/lib/export/exportProfile.ts:50]: cool. let's do that then with the same function as the groups.

File: plugins/@grouparoo/onesignal/src/lib/export/exportProfile.ts (hunk 2 above)
Comment (pedroslopez) [reply plugins/@grouparoo/onesignal/src/lib/export/exportProfile.ts]: @bleonard following this, should we transform booleans to 1/0 in the formatVar?

File: plugins/@grouparoo/onesignal/src/lib/export/exportProfile.ts (hunk 3 above)
Comment (bleonard) [reply plugins/@grouparoo/onesignal/src/lib/export/exportProfile.ts:75]: Let's see if this works with a `id` that is a reproducible UUID from the groupName

File: plugins/@grouparoo/onesignal/src/lib/export/exportProfile.ts (hunk 2 above)
Comment (bleonard) [reply plugins/@grouparoo/onesignal/src/lib/export/exportProfile.ts]: I think they are just trying to save space, so let's not be too surprising for the user and just be "true"

Comment (pedroslopez) [reply plugins/@grouparoo/onesignal/src/lib/export/exportProfile.ts]: @bleonard Same for the in_group tag or keep 1/0 there?

Comment (bleonard) [reply plugins/@grouparoo/onesignal/src/lib/export/exportProfile.ts]: no strong opinion. I guess lets go with "true"

Comment (bleonard) [reply plugins/@grouparoo/onesignal/src/lib/export/exportProfile.ts]: Looks like Segment does true/false https://segment.com/docs/connections/destinations/catalog/onesignal/

**Pull request merged**